### PR TITLE
Property precedence

### LINKED ISSUE #56: Indistinguishable properties and attributes
Currently, the `patch()` function treats every JSX attribute simultaneously as a property *and* an attribute - it initializes both, which means you have no ability to update them individually.

Blindly treating every JSX attribute as both a property and an attribute means that most updates are unnecessarily being done twice - one of them generating either a meaningless property or a meaningless DOM attribute. The DOM accumulates a lot of garbage this way.

For example, `<div innerHTML={content}>` will inject literal HTML via the `innerHTML` property, but will also create a useless `innerHTML` *attribute* - or in some cases, such as `<input value={value}>`, will initialize the value twice, first using the `value` property, then again using `setAttribute('value', ...)`.

I don't have a lot of references, but `snabbdom` for one [separates](https://github.com/snabbdom/snabbdom/blob/master/src/vnode.ts#L23) properties from attributes - which seems sort of inevitable, at the VDOM level, if we want to solve this problem?

### PR DESCRIPTION
**Pull request by mindplay-dk**
Fixes #56

### CODE REVIEW

**Comment by JorgeBucaran** [comment]
@mindplay-dk Now that you understand the issue better (I presume) can you summarize why this change is necessary and what benefit do we get?

**Comment by mindplay-dk** [comment]
The main benefit is we no longer pollute the DOM with meaningless attributes, nor do we waste resources doing so - for example, before, if someone were to set the `className` property, you'd also end up with a meaningless `className` attribute in the DOM.

See also description of #56 for details.

**Comment by mindplay-dk** [comment]
Note that `uglify` currently fails, so this is probably not ready for merge.

```
> uglifyjs dist/picodom.js -o dist/picodom.js --mangle --compress warnings=false --pure-funcs=Object.defineProperty -p relative --in-source-map dist/picodom.js.map --source-map dist/picodom.js.map


Parse error at picodom.js:95,8

    let empty = null == value || false === value;

        ^

SyntaxError: Unexpected token: name (empty)

    at JS_Parse_Error.get (eval at <anonymous> (/home/travis/build/picodom/picodom/node_modules/uglify-js/tools/node.js:27:1), <anonymous>:86:23)

    at /home/travis/build/picodom/picodom/node_modules/uglify-js/bin/uglifyjs:384:39

    at time_it (/home/travis/build/picodom/picodom/node_modules/uglify-js/bin/uglifyjs:620:15)

    at /home/travis/build/picodom/picodom/node_modules/uglify-js/bin/uglifyjs:345:9

    at tryToString (fs.js:449:3)

    at FSReqWrap.readFileAfterClose [as oncomplete] (fs.js:436:12)
```

I've never used `uglify`, but perhaps this is due to the fact I automatically use ES2016 syntax by now? Looks like the tests are targeting an older version of Node?

File: src/patch.js [hunk 1]
```diff
@@ -58,16 +58,18 @@ function setElementProp(element, name, value, oldValue) {
...
-      element[name] = null == value ? "" : value
-    } catch (_) {}
-
-    if (typeof value !== "function") {
-      if (null == value || false === value) {
-        element.removeAttribute(name)
-      } else {
-        element.setAttribute(name, value === true ? "" : value)
-      }
+    let empty = null == value || false === value
```
Comment (JorgeBucaran) [inline src/patch.js]: `var`

**Comment by JorgeBucaran** [comment]
@mindplay-dk Would the following work as well?

```js
function setElementProp(element, name, value, oldValue) {
  if (name === "key") {
  } else if (name === "style") {
    for (var name in merge(oldValue, (value = value || {}))) {
      element.style[name] = value[name] != null ? value[name] : ""
    }
  } else {
    var empty = null == value || false === value

    if (name in element) {
      try {
        element[name] = value == null ? "" : value
      } catch (_) {}
    } else if (typeof value !== "function") {
      if (null == value || false === value) {
        element.removeAttribute(name)
      } else {
        element.setAttribute(name, value === true ? "" : value)
      }
    }
  }
}
```

**Comment by mindplay-dk** [comment]
What's the difference? Just `var` instead of `let`?

That'll work of course. I'm in the habit of using var/let/const by now, it's hard not to 🙄

**Comment by JorgeBucaran** [comment]
@mindplay-dk No, I mean the code I posted above.

**Comment by mindplay-dk** [comment]
Looks functionally equivalent to me, but maybe change it and run the test?

**Comment by JorgeBucaran** [comment]
@mindplay-dk Is there any way we could simplify or compact this code a bit more? It's adding 20 bytes. It seems that setting some properties, like `id`, will cause the attribute to be set automatically, so that's why we need:

```js
if (empty) {
  element.removeAttribute(name)
}
```

...near the end, regardless of previous conditions.

**Comment by mindplay-dk** [comment]
I'm honestly not interested in the file-size - I'm interested in simple, idiomatic, readable code.

If you can do both, that's fine - but "pre-minifying" the code really doesn't help with comprehension.

Isn't there a tool that can aggressively rewrite if/else clauses into compounded ternary expressions and so on? It's really not helpful to anybody - especially given how young this project still is...

Just my two cents.

**Comment by JorgeBucaran** [comment]
@mindplay-dk Not that kind of easy minification, but actualy simplification by rewriting the logic altogether like I did. That didn't really work though, so this looks good. Could you run npm format (or make sure to run prettier on all the files you changed) so I can merge this?

Also, I left you a comment.

File: src/patch.js [hunk 2]
```diff
@@ -58,16 +58,18 @@ function setElementProp(element, name, value, oldValue) {
       element.style[name] = value[name] != null ? value[name] : ""
```
Comment (JorgeBucaran) [inline src/patch.js]: Not part of this PR, but could you change this to `element[name][i]` and also the loop variable from `name` to `i`?

Comment (mindplay-dk) [reply src/patch.js]: Oh, I see, it's overwriting the name variable! Yeah, I'll fix that.

File: src/patch.js [hunk 3]
```diff
@@ -54,20 +49,22 @@ function createElement(node, isSVG) {
 function setElementProp(element, name, value, oldValue) {
   if (name === "key") {
   } else if (name === "style") {
-    for (var name in merge(oldValue, (value = value || {}))) {
-      element.style[name] = value[name] != null ? value[name] : ""
+    for (var i in merge(oldValue, (value = value || {}))) {
+      element.style[i] = value[i] != null ? value[i] : ""
```
Comment (JorgeBucaran) [inline src/patch.js]: @mindplay-dk 
```js
element[name][i]
```

Comment (mindplay-dk) [reply src/patch.js]: Why would you want to do that dynamically? We know it'll only work with `style`, and it's an isolated use-case.

Comment (JorgeBucaran) [reply src/patch.js]: It saves a few bytes.

**Comment by JorgeBucaran** [comment]
Thanks @mindplay-dk! 🎉 😄

File: src/patch.js (hunk 3 above)
Comment (mindplay-dk) [reply src/patch.js]: 2 to be exact 😆

**Comment by JorgeBucaran** [comment]
@mindplay-dk I ported this change to Hyperapp and it broke SVG! 😆 #509
